ddtrace/contrib/trace_utils.py

```diff
@@ -481,10 +482,15 @@ def set_http_meta(
         # We always collect the IP if appsec is enabled to report it on potential vulnerabilities.
         # https://datadoghq.atlassian.net/wiki/spaces/APS/pages/2118779066/Client+IP+addresses+resolution
         if config._appsec_enabled or config.retrieve_client_ip:
-            ip = _get_request_header_client_ip(span, request_headers, peer_ip, headers_are_case_sensitive)
-            if ip:
-                span.set_tag_str(http.CLIENT_IP, ip)
-                span.set_tag_str("network.client.ip", ip)
+            # Retrieve the IP if it was calculated on AppSecProcessor.on_span_start
+            request_ip = _context.get_item("http.request.remote_ip", span=span)
+
+            if not request_ip:
+                # Not calculated: framework does not support IP blocking or testing env
+                request_ip = _get_request_header_client_ip(span, request_headers, peer_ip, headers_are_case_sensitive)
+
+            span.set_tag_str(http.CLIENT_IP, request_ip)
+            span.set_tag_str("network.client.ip", request_ip)
```

Comment:
🤔 I wonder if we should be skipping these if there is no value rather than setting an empty string... but maybe this is a change we should consider for the encoder instead.

Not sure, just still feels weird that we have empty string tag values in the snapshots.

Reply:
Even empty it gives the information that we tried to get the IP but could not, so the user knows he hasn't misconfigured something and it's not retrieving IPs. On the real world this would never be empty since we fallback to the peer IP, that should be always present, if we can't find one in the headers but I prefer it to be empty if IP retrieval is activated either by ASM or the environment var.